pkg/tsdb/prometheus/buffered/time_series_query.go

```diff
@@ -57,41 +57,55 @@ var (
 type Buffered struct {
 	intervalCalculator intervalv2.Calculator
 	tracer             tracing.Tracer
-	getClient          clientGetter
+	client             apiv1.API
 	log                log.Logger
 	ID                 int64
 	URL                string
 	TimeInterval       string
 }
 
-func New(httpClientProvider httpclient.Provider, cfg *setting.Cfg, features featuremgmt.FeatureToggles, tracer tracing.Tracer, settings backend.DataSourceInstanceSettings, plog log.Logger) (*Buffered, error) {
-	var jsonData map[string]interface{}
-	if err := json.Unmarshal(settings.JSONData, &jsonData); err != nil {
-		return nil, fmt.Errorf("error reading settings: %w", err)
+func New(roundTripper http.RoundTripper, tracer tracing.Tracer, settings backend.DataSourceInstanceSettings, plog log.Logger) (*Buffered, error) {
+	promClient, err := CreateClient(roundTripper, settings.URL)
+	if err != nil {
+		return nil, fmt.Errorf("error creating prom client: %v", err)
 	}
 
-	timeInterval, err := maputil.GetStringOptional(jsonData, "timeInterval")
+	jsonData, err := utils.GetJsonData(settings)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error getting jsonData: %w", err)
 	}
 
-	p := promclient.NewProvider(settings, jsonData, httpClientProvider, cfg, features, plog)
-	pc, err := promclient.NewProviderCache(p)
+	timeInterval, err := maputil.GetStringOptional(jsonData, "timeInterval")
 	if err != nil {
 		return nil, err
 	}
+
 	return &Buffered{
 		intervalCalculator: intervalv2.NewCalculator(),
 		tracer:             tracer,
 		log:                plog,
-		getClient:          pc.GetClient,
+		client:             promClient,
 		TimeInterval:       timeInterval,
 		ID:                 settings.ID,
 		URL:                settings.URL,
 	}, nil
 }
 
-func (b *Buffered) runQueries(ctx context.Context, client apiv1.API, queries []*PrometheusQuery) (*backend.QueryDataResponse, error) {
+func (b *Buffered) ExecuteTimeSeriesQuery(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	ctxWithHeaders := sdkHTTPClient.WithContextualMiddleware(ctx, middleware.ReqHeadersMiddleware(req.Headers))
```

Comment:
Adding the middleware here per req

Reply:
I don't think this one is necessary either. It's added in the HTTPClientProvider here:
https://github.com/grafana/grafana/blob/496c2e26f455cc788b31bfa021d8c029803bacd0/pkg/infra/httpclient/httpclientprovider/http_client_provider.go#L29. The problem earlier was that Prometheus datasource somewhere did overwrite the httpclient.Options.Headers.

Query service also adds contextual middlewares for Forwarded OAuth and cookies here:  https://github.com/grafana/grafana/blob/496c2e26f455cc788b31bfa021d8c029803bacd0/pkg/services/query/query.go#L202 

The idea for datasources when using the HTTPClientProvider is to not having to care about custom headers, forwarded OAuth identity and cookies - it should already have been taken care off. Make sense?

Reply:
I think you are right with the custom headers but noticed it later https://github.com/grafana/grafana/pull/51291/files#r904268439. At the same time can we ignore the req.Headers? I mean right now the custom headers are also passed in req.Headers that is why this sort of solved some of the issues even with the bug when creating the client. Aren't there any headers though that should be preserved from front end? Like if in frontend datasource code we manually set some header shouldn't that be propagated?

Reply:
No, the original idea with query headers field was to pass environment info and not any HTTP headers per se. 

With resources it's a bit different since there the idea was to be able to implement your own datasource proxy which is a reverse HTTP proxy.

Reply:
https://github.com/grafana/grafana-plugin-sdk-go/blob/aab437cf0c85570d06c05e6b43e63df967329bf0/proto/backend.proto#L117-L118